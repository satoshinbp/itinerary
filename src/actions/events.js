import db from '../firebase/firebase'

// ADD_EVENT
export const addEvent = event => ({
  type: 'ADD_EVENT',
  event
})

export const startAddEvent = (tripId, eventData = {}) => {
  return dispatch => {
    const {
      title = '',
      date = 0,
      startTime = 0,
      endTime = 0,
      location = '',
      note = ''
    } = eventData
    const event = { title, date, startTime, endTime, location, note }

    return db.collection('trips').doc(tripId).collection('events').add(event).then(snapshot => {
      dispatch(addEvent({
        id: snapshot.id,
        tripId,
        ...event
      }))
    })
  }
}

// EDIT_EVENT
export const editEvent = (id, updates) => ({
  type: 'EDIT_EVENT',
  id,
  updates
})

export const startEditEvent = (id, updates) => {
  return dispatch => {
    return db.collectionGroup('events').get().then(querySnapshot => {
      querySnapshot.forEach(queryDocumentSnapshot => {
        if (queryDocumentSnapshot.id === id) {
          queryDocumentSnapshot.ref.update(updates)
        }
      })
    }).then(() => {
      dispatch(editEvent(id, updates))
    })
  }
}

// REMOVE_EVENT
export const removeEvent = id => ({
  type: 'REMOVE_EVENT',
  id
})

export const startRemoveEvent = (id = '') => {
  return dispatch => {
    return db.collectionGroup('events').get().then(querySnapshot => {
      querySnapshot.forEach(queryDocumentSnapshot => {
        if (queryDocumentSnapshot.id === id) {
          queryDocumentSnapshot.ref.delete()
        }
      })
    }).then(() => {
      dispatch(removeEvent(id))
    })
  }
}

// SET_EVENTS
export const setEvents = events => ({
  type: 'SET_EVENTS',
  events
})

export const startSetEvents = () => {
  return dispatch => {
    return db.collectionGroup('events').get().then(querySnapshot => {
      const events = []

      querySnapshot.forEach(queryDocumentSnapshot => {
        events.push({
          id: queryDocumentSnapshot.id,
          tripId: queryDocumentSnapshot.ref.parent.parent.id,
          ...queryDocumentSnapshot.data()
        })
      })

      dispatch(setEvents(events))
    })
  }
}