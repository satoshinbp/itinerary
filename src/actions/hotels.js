export const addHotel = (
  {
    id='',
    tripId = '',
    name = '',
    checkInDate = 0,
    checkOutDate = 0,
    checkInTime = 0,
    checkOutTime = 0,
    ETA = 0,
    ETD = 0,
    location ='',
    note = ''
  } = {}
) => ({
  type: 'ADD_HOTEL',
  hotel: {
    id,
    tripId,
    name,
    checkInDate,
    checkOutDate,
    checkInTime,
    checkOutTime,
    ETA,
    ETD,
    location,
    note
  }
})

export const editHotel = (id, updates) => ({
  type: 'EDIT_HOTEL',
  id,
  updates
})

export const removeHotel = (id = '') => ({
  type: 'REMOVE_HOTEL',
  id
})