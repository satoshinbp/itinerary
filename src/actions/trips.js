import db from '../firebase/firebase'

// ADD_TRIP
export const addTrip = trip => ({
  type: 'ADD_TRIP',
  trip
})

export const startAddTrip = (tripData = {}) => {
  return dispatch => {
    const {
      title = '',
      startDate = 0,
      endDate = 0,
      note = ''
    } = tripData
    const trip = { title, startDate, endDate, note }

    return db.collection('trips').add(trip).then(snapshot => {
      dispatch(addTrip({
        id: snapshot.id,
        ...trip
      }))
    })
  }
}

// EDIT_TRIP
export const editTrip = (id, updates) => ({
  type: 'EDIT_TRIP',
  id,
  updates
})

export const startEditTrip = (id, updates) => {
  return dispatch => {
    return db.collection('trips').doc(id).update(updates).then(() => {
      dispatch(editTrip(id, updates))
    })
  }
}

// REMOVE_TRIP
export const removeTrip = id => ({
  type: 'REMOVE_TRIP',
  id
})

export const startRemoveTrip = (id = '') => {
  return dispatch => {
    return db.collection('trips').doc(id).delete().then(() => {
      dispatch(removeTrip(id))
    })
  }
}

//SET_TRIPS
export const setTrips = trips => ({
  type: 'SET_TRIPS',
  trips
})

export const startSetTrips = () => {
  return dispatch => {
    return db.collection('trips').get().then(snapshot => {
      const trips = []

      snapshot.forEach(childSnapshot => {
        trips.push({
          id: childSnapshot.id,
          ...childSnapshot.data()
        })
      })

      dispatch(setTrips(trips))
    })
  }
}