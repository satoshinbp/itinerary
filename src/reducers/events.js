const eventsReducerDefaultState = []

export default (state = eventsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EVENT':
      return [
        ...state,
        action.event
      ]
    case 'EDIT_EVENT':
      return state.map((event) => {
        if (event.id === action.id) {
          return {
            ...event,
            ...action.updates
          }
        } else {
          return event
        }
      })
    case 'REMOVE_EVENT':
      return state.filter(({ id }) => id !== action.id)
    default:
      return state
  }
}