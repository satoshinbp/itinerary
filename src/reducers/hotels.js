const hotelsReducerDefaultState = []

export default (state = hotelsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_HOTEL':
      return [
        ...state,
        action.hotel
      ]
    case 'EDIT_HOTEL':
      return state.map((hotel) => {
        if (hotel.id === action.id) {
          return {
            ...hotel,
            ...action.updates
          }
        } else {
          return hotel
        }
      })
    case 'REMOVE_HOTEL':
      return state.filter(({ id }) => id !== action.id)
    default:
      return state
  }
}