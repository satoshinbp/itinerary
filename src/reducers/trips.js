const tripsReducerDefaultState = [];

export default (state = tripsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_TRIP':
      return [
        ...state,
        action.trip
      ];
    case 'EDIT_TRIP':
      return state.map((trip) => {
        if (trip.id === action.id) {
           return {
            ...trip,
            ...action.updates
           }
        } else {
          return trip;
        }
      });
    case 'REMOVE_TRIP':
      return state.filter(({ id }) => id !== action.id);
    default:
      return state;
  }
};