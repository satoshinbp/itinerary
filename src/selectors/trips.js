import moment from 'moment';

export default (trips, { filter }) => {
  return trips.filter((trip) => {
    switch (filter) {
      case 'upcoming':
        return moment(trip.endDate).isSameOrAfter(moment(), 'day');
      case 'past':
        return moment(trip.endDate).isBefore(moment(), 'day');
      case 'all':
        return true;
    }
  }).sort((a, b) => {
    return a.startDate < b.startDate ? 1 : -1;
  });
};