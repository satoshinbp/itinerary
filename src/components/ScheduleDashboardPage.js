import React, { useState } from 'react'
import Modal from 'react-modal'
import { connect } from 'react-redux'
import moment from 'moment'
import ScheduleList from './ScheduleList'
import EventForm from './EventForm'
import HotelForm from './HotelForm'
import { addEvent, editEvent } from '../actions/events'
import { addHotel, editHotel } from '../actions/hotels'

const ScheduleDashboardPage = ({ trip, events, hotels, addEvent, editEvent, addHotel, editHotel }) => {
  const [eventId, setEventId] = useState(undefined)
  const [hotelId, setHotelId] = useState(undefined)
  const [eventDate, setEventDate] = useState(moment())

  const listAllDate = (dates, params_start, params_end) => {
    let start = moment(params_start)

    while (start.diff(params_end, 'day') <= 0) {
      if (!dates.find(date => date.isSame(start, 'day'))) {
        dates.push(moment(start))
      }
      start.add(1, 'day')
    }
  }

  let dates = []
  listAllDate(dates, trip.startDate, trip.endDate)
  events.map(event => {
    if (!dates.find(date => date.isSame(event.date, 'day'))) {
      dates.push(moment(event.date))
    }
  })
  hotels.map(hotel => listAllDate(dates, hotel.checkInDate, hotel.checkOutDate))
  dates.sort((a, b) => a.isAfter(b) ? 1 : -1)

  return (
    <div className="container">
      <div className="page-header__title">
        <h3>{trip.title}</h3>
        <p>
          {
            trip.startDate.isSame(trip.endDate, 'day') ? (
              `${trip.startDate.format('YYYY/MM/DD')}`
            ) : (
                `${trip.startDate.format('YYYY/MM/DD')} - ${trip.endDate.format('YYYY/MM/DD')}`
              )
          }
        </p>
      </div>
      {
        dates.map((date, index) => (
          <ScheduleList
            key={index}
            date={date}
            tripId={trip.id}
            setEventId={setEventId}
            setHotelId={setHotelId}
            setEventDate={setEventDate}
          />
        ))
      }
      <Modal
        className="modal"
        isOpen={!!eventId || !!hotelId}
        contentLabel="Event/Hotel Form"
        onRequestClose={() => {
          setEventId(undefined)
          setHotelId(undefined)
          setEventDate(moment())
        }}
        closeTimeoutMS={200}
      >
        {
          !!eventId ? (
            <EventForm
              onSubmit={event => {
                if (events.find(event => event.id === eventId)) {
                  editEvent(eventId, event)
                } else {
                  addEvent(event)
                }
                setEventId(undefined)
                setEventDate(moment())
              }}
              id={eventId}
              tripId={trip.id}
              date={eventDate}
              event={events.find(event => event.id === eventId)}
            />
          ) : (
              <HotelForm
                onSubmit={hotel => {
                  if (hotels.find(hotel => hotel.id === hotelId)) {
                    editHotel(hotelId, hotel)
                  } else {
                    addHotel(hotel)
                  }
                  setHotelId(undefined)
                  setEventDate(moment())
                }}
                id={hotelId}
                tripId={trip.id}
                date={eventDate}
                hotel={hotels.find(hotel => hotel.id === hotelId)}
              />
            )
        }
      </Modal>
    </div>
  )
}

const mapStateToProps = (state, props) => ({
  trip: state.trips.find(trip => trip.id === props.match.params.id),
  events: state.events.filter(event => event.tripId === props.match.params.id),
  hotels: state.hotels.filter(hotel => hotel.tripId === props.match.params.id)
})

const mapDispatchToProps = dispatch => ({
  addEvent: event => dispatch(addEvent(event)),
  editEvent: (id, updates) => dispatch(editEvent(id, updates)),
  addHotel: hotel => dispatch(addHotel(hotel)),
  editHotel: (id, updates) => dispatch(editHotel(id, updates))
})

export default connect(mapStateToProps, mapDispatchToProps)(ScheduleDashboardPage)