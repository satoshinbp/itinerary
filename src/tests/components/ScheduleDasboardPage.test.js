import React from 'react'
import { shallow } from 'enzyme'
import trips from '../fixtures/trips'
import events from '../fixtures/events'
import hotels from '../fixtures/hotels'
import { ScheduleDashboardPage } from '../../components/ScheduleDashboardPage'

test('should render ScheduleDashboardPage correctly for one day trip', () => {
  const wrapper = shallow(<ScheduleDashboardPage trip={trips[2]} events={[events[2]]} hotels={[]} />)
  expect(wrapper).toMatchSnapshot()
})

test('should render ScheduleDashboardPage correctly for multi days trip', () => {
  const wrapper = shallow(<ScheduleDashboardPage trip={trips[0]} events={[events[0]]} hotels={[hotels[0], hotels[2]]} />)
  expect(wrapper).toMatchSnapshot()
})


// test for Modal open case to be implemented