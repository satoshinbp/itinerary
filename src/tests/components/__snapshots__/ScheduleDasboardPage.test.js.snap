// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ScheduleDashboardPage correctly for multi days trip 1`] = `
<div
  className="container"
>
  <div
    className="page-header__title"
  >
    <h3>
      Summer Vacation in Hawaii
    </h3>
    <p>
      1970/01/01 - 1970/01/03
    </p>
  </div>
  <Connect(ScheduleList)
    date={"1970-01-01T00:00:00.000Z"}
    key="0"
    setDateOnEdit={[Function]}
    setEventId={[Function]}
    setHotelId={[Function]}
    tripId="a"
  />
  <Connect(ScheduleList)
    date={"1970-01-02T00:00:00.000Z"}
    key="1"
    setDateOnEdit={[Function]}
    setEventId={[Function]}
    setHotelId={[Function]}
    tripId="a"
  />
  <Connect(ScheduleList)
    date={"1970-01-03T00:00:00.000Z"}
    key="2"
    setDateOnEdit={[Function]}
    setEventId={[Function]}
    setHotelId={[Function]}
    tripId="a"
  />
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="modal"
    closeTimeoutMS={200}
    contentLabel="Event/Hotel Form"
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <_default
      date={"1970-01-01T00:00:00.000Z"}
      onSubmit={[Function]}
      tripId="a"
    />
  </Modal>
</div>
`;

exports[`should render ScheduleDashboardPage correctly for one day trip 1`] = `
<div
  className="container"
>
  <div
    className="page-header__title"
  >
    <h3>
      Tokyo - 2020 Spring
    </h3>
    <p>
      1970/01/11
    </p>
  </div>
  <Connect(ScheduleList)
    date={"1970-01-11T00:00:00.000Z"}
    key="0"
    setDateOnEdit={[Function]}
    setEventId={[Function]}
    setHotelId={[Function]}
    tripId="c"
  />
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="modal"
    closeTimeoutMS={200}
    contentLabel="Event/Hotel Form"
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <_default
      date={"1970-01-01T00:00:00.000Z"}
      onSubmit={[Function]}
      tripId="c"
    />
  </Modal>
</div>
`;
